File: server/src/main/java/com/nexoffer/controller/SkillGapController.java
package com.nexoffer.controller;

import com.nexoffer.model.History;
import com.nexoffer.model.Profile;
import com.nexoffer.model.User;
import com.nexoffer.repository.HistoryRepository;
import com.nexoffer.repository.ProfileRepository;
import com.nexoffer.service.LlmGatewayService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/skill-gap")
public class SkillGapController {

    private final ProfileRepository profileRepository;
    private final HistoryRepository historyRepository;
    private final LlmGatewayService llmGatewayService;

    public SkillGapController(ProfileRepository profileRepository,
                              HistoryRepository historyRepository,
                              LlmGatewayService llmGatewayService) {
        this.profileRepository = profileRepository;
        this.historyRepository = historyRepository;
        this.llmGatewayService = llmGatewayService;
    }

    @PostMapping("/analyze")
    public ResponseEntity<?> analyzeSkillGap(@AuthenticationPrincipal User user, @RequestBody(required = false) Map<String, String> body) {
        String resumeText = body != null ? body.get("resumeText") : null;
        String jobDescription = body != null ? body.get("jobDescription") : null;
        String company = body != null ? body.get("company") : null;
        String role = body != null ? body.get("role") : null;

        Optional<Profile> profileOpt = profileRepository.findByUserId(user.getId());
        if (profileOpt.isPresent()) {
            Profile p = profileOpt.get();
            if (resumeText == null || resumeText.isBlank()) resumeText = p.getResumeText();
            if (jobDescription == null || jobDescription.isBlank()) jobDescription = p.getJobDescription();
            if (company == null || company.isBlank()) company = p.getCompany();
            if (role == null || role.isBlank()) role = p.getRole();
        }

        if (resumeText == null || resumeText.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("success", false, "message", "Resume text is missing. Please upload your resume in Profile first."));
        }

        String targetCompany = (company != null && !company.isBlank()) ? company : "Target Company";
        String targetRole = (role != null && !role.isBlank()) ? role : "Target Role";

        String prompt = "Compare the candidate's Resume against the target Job Description (for " + targetRole + " at " + targetCompany + ").\n" +
                "Conduct a comprehensive Skill Gap Analysis.\n" +
                "Classify skills into existingSkills, missingSkills (with priority: High|Medium|Low, reason, recommendedAction), and preparationSummary.\n" +
                "Return ONLY valid JSON with this schema:\n" +
                "{\n" +
                "  \"existingSkills\": [{ \"skill\": \"...\", \"proficiency\": \"...\", \"notes\": \"...\" }],\n" +
                "  \"missingSkills\": [{ \"skill\": \"...\", \"priority\": \"High|Medium|Low\", \"reason\": \"...\", \"recommendedAction\": \"...\" }],\n" +
                "  \"preparationSummary\": \"...\"\n" +
                "}";

        Map<String, String> context = new HashMap<>();
        context.put("resumeText", resumeText);
        context.put("jobDescription", jobDescription != null ? jobDescription : "Standard tech requirements");
        context.put("company", targetCompany);
        context.put("role", targetRole);

        LlmGatewayService.AiResponse aiResponse = llmGatewayService.generateAIResponse(
                prompt,
                "You are an expert Technical Skills Assessor and Career Transition Coach.",
                context,
                true,
                "skill-gap"
        );

        Object data = aiResponse.getParsedJson() != null ? aiResponse.getParsedJson() : llmGatewayService.generateFallbackResponse("skill-gap", context);

        historyRepository.save(new History(
                user.getId(),
                "Skill Gap Analyzer",
                targetCompany,
                targetRole,
                "Identified skill gaps against target JD"
        ));

        return ResponseEntity.ok(Map.of(
                "success", true,
                "data", data,
                "provider", aiResponse.getProvider()
        ));
    }
}

File: server/controllers/skillGapController.js
const { ProfileRepository, HistoryRepository } = require('../models/dbStore');
const llmGateway = require('../services/llmGateway');

// @desc    Analyze skill gaps between resume and JD
// @route   POST /api/skill-gap/analyze
const analyzeSkillGap = async (req, res, next) => {
  try {
    let { resumeText, jobDescription, company, role } = req.body;

    const profile = await ProfileRepository.findOne({ userId: req.user._id });
    if (profile) {
      if (!resumeText) resumeText = profile.resumeText;
      if (!jobDescription) jobDescription = profile.jobDescription;
      if (!company) company = profile.company;
      if (!role) role = profile.role;
    }

    if (!resumeText) {
      return res.status(400).json({
        success: false,
        message: 'Resume text is missing. Please upload your resume in Profile first.',
      });
    }

    const prompt = `
Compare the candidate's Resume against the target Job Description (for ${role || 'Target Role'} at ${company || 'Target Company'}).
Conduct a comprehensive Skill Gap Analysis.
Classify skills into:
1. "existingSkills": List of skills found in the resume with assessed proficiency (Beginner|Intermediate|Advanced) and brief notes.
2. "missingSkills": List of skills demanded by the JD that are not found or weak in the resume. For each missing skill, assign:
   - "priority": "High" | "Medium" | "Low"
   - "reason": Why it is critical for this role
   - "recommendedAction": Practical step to bridge this gap quickly.
3. "preparationSummary": Concise action plan.

Return ONLY valid JSON with this schema:
{
  "existingSkills": [
    { "skill": "...", "proficiency": "...", "notes": "..." }
  ],
  "missingSkills": [
    { "skill": "...", "priority": "High|Medium|Low", "reason": "...", "recommendedAction": "..." }
  ],
  "preparationSummary": "..."
}
`;

    const context = {
      resumeText,
      jobDescription: jobDescription || 'Standard requirements for modern software and tech roles.',
      company,
      role,
    };

    const aiResult = await llmGateway.generateAIResponse({
      prompt,
      context,
      jsonMode: true,
      moduleType: 'skill-gap',
      systemPrompt: 'You are an expert Technical Skills Assessor and Career Transition Coach.',
    });

    const data = aiResult.parsedJson || llmGateway.generateFallbackResponse('skill-gap', context);

    await HistoryRepository.create({
      userId: req.user._id,
      module: 'Skill Gap Analyzer',
      company: company || 'General',
      role: role || 'General',
      summary: `Found ${data.existingSkills?.length || 0} existing skills and ${data.missingSkills?.length || 0} skill gaps`,
      metadata: { missingCount: data.missingSkills?.length || 0 },
    });

    res.status(200).json({
      success: true,
      data,
      provider: aiResult.provider,
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  analyzeSkillGap,
};
